core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -684,6 +690,32 @@ Cuda::size_type *cuda_internal_scratch_unified(const Cuda &instance,
 
 namespace Kokkos {
 
+bool Cuda::is_running() const {
+  switch (m_space_instance->m_internal_status) {
+    case Impl::ExecutionSpaceStatus::complete: return false;
+    case Impl::ExecutionSpaceStatus::submitted:
+      KOKKOS_IMPL_CUDA_SAFE_CALL(cudaEventRecord(m_space_instance->m_last_event,
```

Comment:
Is `cudaEventRecord` even thread-safe with respect to a common stream and event?

I could readily see even `cudaEventQuery` not being thread-safe, if it acts at all like an MPI progress engine in a call to `MPI_Wait` - i.e. actually advancing internal state.

Reply:
With the added lock, this is no longer a concern.